Fullscreen support

emersion opened the pull request:
This PR implements basic fullscreen support. This PR doesn't implement hardware fullscreen, this will be done in a separate PR. We need to do software fullscreen anyway in case hardware fullscreen is not supported.

- [x] xdg-shell support
- [x] wl-shell support
- [x] xwayland support
- [x] Render subsurfaces/popups/...
- [ ] ~~Render subsurfaces in `wlr_output`?~~ No compositing in `wlr_output`

Test plan:
- [x] `mpv`
  - [x] xwayland
  - [x] wl-shell
- [x] `evince`
  - [x] xdg-shell
  - [x] xwayland
- [x] `epiphany` (xdg-shell)
- [x] `chromium` (xwayland)
- [x] Test scaled outputs
- [x] Test rotated outputs

SirCmpwn commented:
We do have to render subsurfaces.

emersion commented:
All right, this is ready for review!

SirCmpwn commented:
Can you add a keybinding to fullscreen arbitrary windows?

acrisci commented on include/rootston/view.h:
```diff
@@ -13,6 +13,7 @@ struct roots_wl_shell_surface {
 	struct wl_listener request_move;
 	struct wl_listener request_resize;
 	struct wl_listener request_set_maximized;
+	struct wl_listener request_set_fullscreen;
```
Naming should be consistent across the views if possible (so `request_fullscreen` and `request_mazimized`).

emersion replied:
I tried to respect each shell's terminology. I'll unify all of them then.

acrisci commented on include/wlr/types/wlr_xdg_shell_v6.h:
```diff
@@ -152,6 +152,13 @@ struct wlr_xdg_toplevel_v6_resize_event {
 };
 
+struct wlr_xdg_toplevel_v6_set_fullscreen_event {
+	struct wl_client *client;
```
Why are we putting the client in these events? Shouldn't it be always the same as the client for the shell surface?

emersion replied:
I just copy-pasta-ed the boilerplate from other events. So yeah, I don't think the client is useful in these events. Let's remove it.

ascent12 commented on types/wlr_surface.c:
```diff
@@ -929,3 +929,17 @@ void wlr_surface_send_leave(struct wlr_surface *surface,
 	}
 }
+
+static inline int64_t timespec_to_msec(const struct timespec *a) {
```
We have this functions in a couple of places.
It should probably live in util or something.

emersion replied:
Actually, there's only one of them in wlroots, which is this one. All others are in `examples/` (in `output-layout.c`, `rotation.c` and `simple.c`), so I wonder if it's worth it.

acrisci commented on rootston/wl_shell.c:
```diff
@@ -59,6 +59,17 @@ static void handle_request_set_maximized(struct wl_listener *listener,
+	struct roots_view *view = roots_surface->view;
+	struct wlr_wl_shell_surface_set_fullscreen_event *e = data;
+
+	// TODO: support e->method, e->framerate
```
I really doubt we'll ever support these things.

acrisci commented:
Why couldn't we do direct scanout for a surface with subsurfaces?

emersion commented:
If I've understood correctly, the goal of direct scan-out is avoiding a rendering step: the buffer is directly displayed in an overlay plane (maybe even without a copy, from what I've seen from weston's source). Correct me if I'm wrong, I'm not familiar with this.

SirCmpwn commented:
Same reason we can't do it for all views at all times. They need to be composited.

acrisci commented on types/wlr_xdg_shell_v6.c:
```diff
@@ -677,15 +663,38 @@ static void xdg_toplevel_protocol_unset_maximized(struct wl_client *client,
+		.output = NULL,
+	};
+
+	wl_signal_emit(&surface->events.request_fullscreen, &event);
```
I think we wait until the shell surface commits the whole state before we emit the event.

emersion replied:
What do you mean? We should wait? Right now `request_fullscreen` isn't triggered on commit.

Note: the state doesn't keep track of the `output` parameter, which would make waiting for a commit more tricky.

acrisci replied:
I'm making a note of this now because I'm pretty sure from looking at the code we're going to see rendering artifacts when the fullscreen happens. I'll test it when you make the command.

acrisci commented on types/wlr_output.c:
```diff
@@ -252,6 +252,27 @@ void wlr_output_make_current(struct wlr_output *output) {
 }
 
+static void output_fullscreen_surface_render(struct wlr_output *output,
+		struct wlr_surface *surface, const struct timespec *when) {
```
Is the whole point of handling the rendering here to support direct scanout later?

emersion replied:
Yes. This is a software fallback in case overlay planes and direct scan-out isn't supported.

emersion commented:
>Can you add a keybinding to fullscreen arbitrary windows?

Added the `fullscreen` command that toggles fullscreen for the focused view. Some views don't expect that (e.g. `gnome-calculator` is buggy, have to click outside the view when fullscreened otherwise it's clipped) but most of them react correctly.

Addressed all other comments.

acrisci commented:
I'm getting lots of bugs with this.

1. partially rendered content on fullscreen similar to #439 
2. gets the wrong size on my output (drm and x11 backend)
3. pointer is messed up after return to normal state - doesn't respond to click events and pointer is put in the wrong location on the view (probably related to client cursors)

emersion commented:
Yeah, this is pretty strange. `wlr_output_set_fullscreen_surface` messes up the output, even after there's no fullscreen surface anymore. I'll look into this.

I'd like to address the partially rendered content in another PR, solving the issue at the same time for fullscreen and maximize.

emersion commented:
@acrisci Issues 2 and 3 should be fixed with the last commit.

emersion commented:
https://github.com/swaywm/wlroots/pull/437/commits/3262661e1e77db11d7e4939aa377804046df43f5 fixes a bug: `wl_output.enter` wasn't sent to the output at (0, 0) in the layout. This fix reveals a bug: on rotated outputs, `weston-terminal` draws an already rotated surface, and we rotate it a second time. I'll open an issue when this PR gets merged.

acrisci commented:
I'm still seeing a little jump sometimes on the transition. The window moves around before it goes in and out of fullscreen.

Also, take the window in and out of fullscreen many times and you will see that the size shrinks every time.

emersion commented:
>I'm still seeing a little jump sometimes on the transition.

As I said, I'll fix this with #439 in another PR since the fixes for these two bugs will be very similar.

>Also, take the window in and out of fullscreen many times and you will see that the size shrinks every time.

Which client? I can't reproduce with `evince` or `chromium`.

acrisci commented:
I can't reproduce it with those either. It's just gnome-terminal.

emersion commented:
>I can't reproduce it with those either. It's just gnome-terminal.

I don't think we can do anything about it. I can reproduce in GNOME.

acrisci commented:
👍
